### PR TITLE
Idris hackathon

### PR DESCRIPTION
**Pull request by pierre-weiss**
### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
This does not work. Many examples are broken, probably due to the trainer changes.

- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: examples/patch-priors/demo_epll.py [hunk 1]
```diff
@@ -27,7 +26,8 @@
 url = get_image_url("CBSD_0010.png")
 test_img = load_url_image(url, grayscale=False).to(device)
 patch_size = 6
-model = EPLL(channels=test_img.shape[1], patch_size=patch_size, device=device)
+with torch.no_grad():
```
Comment (matthieutrs) [inline examples/patch-priors/demo_epll.py:29]: why is this needed?

Comment (pierre-weiss) [reply examples/patch-priors/demo_epll.py:29]: When we ran the make html yesterday, this demo (and others) saturated the GPU memory. 
This does not happen with torch.nograd.

Comment (matthieutrs) [reply examples/patch-priors/demo_epll.py:29]: ok, but unless I'm mistaken the one below should be sufficient?

File: deepinv/physics/functional/hist.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    :param float low: The lower bound. If `low` is `None` the min of `x` is used instead.
+    :param float upp: The upper bound. If `upp` is `None` the max of `x` is used instead.
+    :param kwargs: Keyword arguments passed to `histogramdd`.
+
+    :return torch.Tensor: The histogram
+    """
+
+    return histogramdd(x.unsqueeze(-1), bins, low, upp, **kwargs)
+
+# if __name__ == "__main__":  # bad practice
```
Comment (matthieutrs) [inline deepinv/physics/functional/hist.py:231]: Can we remove?

File: deepinv/physics/functional/interp.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        self._fitted = False
+        self.alpha = alpha
+        self.device = torch.device(device)
+        self.dtype = dtype
+
+        self.parameters = torch.tensor([], dtype=torch.float32)
+        self.control_points = torch.tensor([], dtype=torch.float32)
+
+    def fit(self, X: torch.Tensor, Y: torch.Tensor):
+        """Learn f that matches Y given X
```
Comment (matthieutrs) [inline deepinv/physics/functional/interp.py:88]:
```suggestion
        r"""Learn f that matches Y given X
```

File: deepinv/physics/functional/interp.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+        Y = torch.vstack([Y, torch.zeros((d_s + 1, Y.shape[1]), device=self.device)])
+
+        self.parameters = torch.linalg.solve(A, Y)  # pylint: disable=not-callable
+        self._fitted = True
+
+        return self
+
+    def transform(self, X: torch.Tensor) -> torch.Tensor:
+        """Map source space to target space
```
Comment (matthieutrs) [inline deepinv/physics/functional/interp.py:138]:
```suggestion
        r"""Map source space to target space
```

File: deepinv/physics/functional/interp.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+        phi = self._radial_distance(X)  # n x n_c
+
+        X = torch.hstack(
+            [phi, torch.ones((X.shape[0], 1), device=self.device), X]
+        )  # n x (n_c + 1 + d_s)
+        return X @ self.parameters
+
+    def _radial_distance(self, X: torch.Tensor) -> torch.Tensor:
+        """Compute the pairwise radial distances of the given points to the control points
```
Comment (matthieutrs) [inline deepinv/physics/functional/interp.py:162]:
```suggestion
        r"""Compute the pairwise radial distances of the given points to the control points
```

File: deepinv/physics/functional/interp.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        # Don't use mm for euclid dist, lots of imprecision comes from it (Will be a bit slower)
+        dist = torch.cdist(
+            X, self.control_points, compute_mode="donot_use_mm_for_euclid_dist"
+        )
+        dist[dist == 0] = 1  # phi(r) = r^2 log(r) ->  (phi(0) = 0)
+        return dist**2 * torch.log(dist)
+
+
+def _ensure_2d(tensor: torch.Tensor) -> torch.Tensor:
+    """Ensure that tensor is a 2d tensor
```
Comment (matthieutrs) [inline deepinv/physics/functional/interp.py:183]:
```suggestion
    r"""Ensure that tensor is a 2d tensor
```

File: deepinv/physics/functional/radon.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    p2 = torch.arange(-(n // 2), 0)
+    results[N:] = p2
+    return results * val
+
+
+def deg2rad(x):
+    return x * 4 * torch.ones(1, device=x.device, dtype=x.dtype).atan() / 180
+
+
+class AbstractFilter(nn.Module):
```
Comment (matthieutrs) [inline deepinv/physics/functional/radon.py:35]: TODO: add docstrings

File: deepinv/physics/functional/radon.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+            R = torch.tensor(
+                [[[theta.cos(), theta.sin(), 0], [-theta.sin(), theta.cos(), 0]]],
+                dtype=self.dtype,
+                device=device,
+            )
+            all_grids.append(affine_grid(R, torch.Size([1, 1, grid_size, grid_size])))
+        return torch.stack(all_grids)
+
+
+class IRadon(nn.Module):
```
Comment (matthieutrs) [inline deepinv/physics/functional/radon.py:153]: TODO: add docstrings

File: deepinv/physics/generator/base.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import torch.nn as nn
+from typing import List
+
+
+class PhysicsGenerator(nn.Module):
+    r"""
+    Base class for parameter generation of physics parameters.
+
+    PhysicsGenerators can be summed to create larger generators (see :meth:`deepinv.physics.generator.PhysicsGenerator.__add__`),
+    or mixed to create a generator that randomly selects (see :meth:`deepinv.physics.generator.GeneratorMixture`).
```
Comment (matthieutrs) [inline deepinv/physics/generator/base.py:11]: Is it possible to document the input args / returns, and to add an example of how the step works in the below example?

File: deepinv/physics/generator/blur.py [hunk 10]
```diff
@@ -0,0 +1,673 @@
...
+
+    def matern_kernel(self, diff, sigma: float = None, l: float = None):
+        if sigma is None:
+            sigma = self.sigma
+        if l is None:
+            l = self.l
+        fraction = 5**0.5 * diff.abs() / l
+        return sigma**2 * (1 + fraction + fraction**2 / 3) * torch.exp(-fraction)
+
+    # @torch.compile
```
Comment (matthieutrs) [inline deepinv/physics/generator/blur.py:99]: do we keep the decorator?

File: deepinv/physics/generator/blur.py [hunk 11]
```diff
@@ -0,0 +1,673 @@
...
+
+        :return: dictionary with key **'filter'**: the generated PSF of shape `(batch_size, 1, psf_size[0], psf_size[1])`
+        :rtype: dict
+        """
+        ## add batch size to the shape. We can have a different batch size at each call of step()
+        ## We enforce only one channel as the underlying generator code only works for one channel at the time
+        # if self.shape[0] != 1:
+        #    self.shape = (batch_size, 1, self.shape[-2], self.shape[-1])
+        # else:
+        #    self.shape = (batch_size, self.shape[-3], self.shape[-2], self.shape[-1])
```
Comment (matthieutrs) [inline deepinv/physics/generator/blur.py:132]: remove?

File: deepinv/physics/generator/blur.py [hunk 12]
```diff
@@ -0,0 +1,673 @@
...
+        self.Z = torch.zeros(
+            (self.pupil_size[0], self.pupil_size[1], self.n_zernike),
+            **self.factory_kwargs,
+        )
+        for k in range(len(self.index_params)):
+            self.Z[:, :, k] = list_zernike_polynomial[self.index_params[k]](
+                XX, YY
+            )  # defining the k-th Zernike polynomial
+
+    def __update__(self):
```
Comment (matthieutrs) [inline deepinv/physics/generator/blur.py:273]: Should we add docstrings?

File: deepinv/physics/generator/blur.py [hunk 13]
```diff
@@ -0,0 +1,673 @@
...
+                self.psf_size[1],
+            )
+        }
+
+
+class DiffractionBlurGenerator(PSFGenerator):
+    r"""
+    Diffraction limited blur generator.
+
+    Generates 2D diffraction kernels in optics using Zernike decomposition of the phase mask (Fresnel/Fraunhoffer diffraction theory)
```
Comment (matthieutrs) [inline deepinv/physics/generator/blur.py:167]:
```suggestion
    Generates 2D diffraction kernels in optics using Zernike decomposition of the phase mask (Fresnel/Fraunhoffer diffraction theory).
```

This is great! Maybe add a ref?

File: deepinv/physics/generator/blur.py [hunk 14]
```diff
@@ -0,0 +1,673 @@
...
+        ].unsqueeze(1)
+
+        psf = psf3 / torch.sum(psf3, dim=(-1, -2), keepdim=True)
+        return {
+            "filter": psf.expand(-1, self.shape[0], -1, -1),
+            "coeff": coeff,
+            "pupil": pupil3,
+        }
+
+    def generate_coeff(self, batch_size):
```
Comment (matthieutrs) [inline deepinv/physics/generator/blur.py:314]: Is it possible to add docstrings?

File: deepinv/physics/generator/blur.py [hunk 15]
```diff
@@ -0,0 +1,673 @@
...
+        coeff = torch.rand((batch_size, len(self.list_param)), **self.factory_kwargs)
+        coeff = (coeff - 0.5) * self.max_zernike_amplitude
+        return coeff
+
+
+def define_zernike():
+    r"""
+    Returns a list of Zernike polynomials lambda functions in Cartesian coordinates
```
Comment (matthieutrs) [inline deepinv/physics/generator/blur.py:322]:
```suggestion
    Returns a list of Zernike polynomials lambda functions in Cartesian coordinates.
```

File: deepinv/physics/generator/blur.py [hunk 16]
```diff
@@ -0,0 +1,673 @@
...
+
+    :param list[func]: list of 37 lambda functions with the Zernike Polynomials. They are ordered as follows
```
Comment (matthieutrs) [inline deepinv/physics/generator/blur.py:324]:
```suggestion
    :param list[func]: list of 37 lambda functions with the Zernike Polynomials. They are ordered as follows:
```

File: deepinv/physics/generator/blur.py [hunk 17]
```diff
@@ -0,0 +1,673 @@
...
+
+    :param torch.Tensor x: x coordinates
+    :param torch.Tensor y: y coordinates
+
+    :return: tuple (rho, phi) of torch.Tensor with radius and angle
+    :rtype: tuple
+    """
+
+    rho = torch.sqrt(x**2 + y**2)
+    # phi = torch.arctan2(y, x)
```
Comment (matthieutrs) [inline deepinv/physics/generator/blur.py:486]:
```suggestion
```

File: deepinv/physics/generator/blur.py [hunk 18]
```diff
@@ -0,0 +1,673 @@
...
+        w = tps.transform(torch.stack((yy.flatten(), xx.flatten()), dim=1)).T
+        w = w.reshape(self.n_eigen_psf, n0, n1)[:, None, None]
+
+        # Ending
+        params_blur = {"h": eigen_psf, "w": w, "padding": self.padding}
+        return params_blur
+
+
+# # %%
+# if __name__ == "__main__":
```
Comment (matthieutrs) [inline deepinv/physics/generator/blur.py:605]: Can we remove this?

File: deepinv/physics/generator/mri.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+            )
+            mask[i, :, random_line_indices] = 1
+
+        # for i in range(batch_size):
+        #     random_line_indices = torch.multinomial(
+        #         torch.arange(img_size[0], **self.factory_kwargs),
+        #         num_samples=num_lines_side // 2,
+        #         replacement=False,
+        #     )
+        #     mask[i, :, random_line_indices] = 1
```
Comment (matthieutrs) [inline deepinv/physics/generator/mri.py:79]:
```suggestion
```

(is it ok to remove?)

File: deepinv/physics/generator/mri.py [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+        return {"mask": torch.cat([mask.float().unsqueeze(1)] * 2, dim=1)}
+
+
+# if __name__ == "__main__":
```
Comment (matthieutrs) [inline deepinv/physics/generator/mri.py:83]: Can we remove?

**Review by matthieutrs — CHANGES_REQUESTED** [review]
Thanks for the great PR 👏  

I'm only half way through but here is a first batch of comments. Happy to give a hand if needed!

File: deepinv/optim/utils.py [hunk 21]
```diff
@@ -39,30 +51,26 @@ def conjugate_gradient(A, b, max_iter=1e2, tol=1e-5):
     :param torch.Tensor b: input tensor
     :param int max_iter: maximum number of CG iterations
     :param float tol: absolute tolerance for stopping the CG algorithm.
+    :param float eps: a small value for numerical stability
```
[the remark this answers is not in the document]
Comment (minh-hai-nguyen-712) [reply deepinv/optim/utils.py:54]: Added!

File: deepinv/optim/utils.py [hunk 22]
```diff
@@ -27,7 +28,18 @@ def check_conv(X_prev, X, it, crit_conv="residual", thres_conv=1e-3, verbose=Fal
         return False
 
 
-def conjugate_gradient(A, b, max_iter=1e2, tol=1e-5):
+def dot(a, b):
+    dot = (a.conj() * b).sum(dim=(-1, -2, -3), keepdim=True)
```
[the remark this answers is not in the document]
Comment (minh-hai-nguyen-712) [reply deepinv/optim/utils.py:32]: Thanks for the doc!

File: docs/source/deepinv.physics.rst [hunk 23]
```diff
@@ -1,111 +1,380 @@
 .. _physics:
 
 Physics
-=======
+=========
```
Comment (matthieutrs) [inline docs/source/deepinv.physics.rst:4]:
```suggestion
=======
```

File: docs/source/deepinv.physics.rst [hunk 24]
```diff
@@ -1,111 +1,380 @@
...
+
+.. doctest::
+
+   >>> import torch
+   >>> import deepinv as dinv
+   >>>
+   >>> from deepinv.physics import Blur
+   >>> x = torch.rand((1, 1, 16, 16))
+   >>> theta = torch.ones((1, 1, 2, 2)) / 4 # a basic 2x2 averaging filter
+   >>> # %% A first possibility
```
Comment (matthieutrs) [inline docs/source/deepinv.physics.rst:56]:
```suggestion
   >>> # A first possibility
```

File: docs/source/deepinv.physics.rst [hunk 25]
```diff
@@ -1,111 +1,380 @@
...
+   >>> physics = Blur(filter=theta) # we instantiate a blur operator with its convolution filter
+   >>> y = physics(x)
+   >>> 
+   >>> # %% A second possibilty
```
Comment (matthieutrs) [inline docs/source/deepinv.physics.rst:60]:
```suggestion
   >>> # A second possibilty
```

File: docs/source/deepinv.physics.rst [hunk 26]
```diff
@@ -1,111 +1,380 @@
...
+   >>> physics = Blur() # a blur operator without convolution filter
+   >>> y = physics(x, filter=theta) # we define the blur by specifying its filter 
+   >>> y = physics(x) # now, the filter is well defined and this line does the same as above 
+   >>> y = physics(x, filter=torch.rand((1,1,2,2))) # we set and apply a random blur filter 
+   >>> y = physics(x) # the convolution filter is now random 
+   >>> 
+   >>> # %% The same can be done by passign in a dictionary including 'filter' as a key
```
Comment (matthieutrs) [inline docs/source/deepinv.physics.rst:67]:
```suggestion
   >>> # The same can be done by passign in a dictionary including 'filter' as a key 
```

File: docs/source/deepinv.physics.rst [hunk 27]
```diff
@@ -1,111 +1,380 @@
...
+
+   >>> import torch
+   >>> import deepinv as dinv
+   >>>
+   >>> from deepinv.physics import Blur
+   >>> x = torch.rand((1, 1, 16, 16))
+   >>> theta = torch.ones((1, 1, 2, 2)) / 4 # a basic 2x2 averaging filter
+   >>> # %% A first possibility
+   >>> physics = Blur(filter=theta) # we instantiate a blur operator with its convolution filter
+   >>> y = physics(x)
```
Comment (matthieutrs) [inline docs/source/deepinv.physics.rst:58]:
```suggestion
   >>> y = physics.A(x)
```

I think it is important to make a distinction between this section where we explain how to apply an operator (that is `A`, i.e. `physics.A(...)`, which is different from `physics(...)` that contains that noise.

File: docs/source/deepinv.physics.rst [hunk 28]
```diff
@@ -1,111 +1,380 @@
...
+   >>> 
+   >>> # %% A second possibilty
+   >>> physics = Blur() # a blur operator without convolution filter
+   >>> y = physics(x, filter=theta) # we define the blur by specifying its filter 
+   >>> y = physics(x) # now, the filter is well defined and this line does the same as above 
+   >>> y = physics(x, filter=torch.rand((1,1,2,2))) # we set and apply a random blur filter 
+   >>> y = physics(x) # the convolution filter is now random
```
Comment (matthieutrs) [inline docs/source/deepinv.physics.rst:65]:
```suggestion
   >>> y = physics.A(x, filter=theta) # we define the blur by specifying its filter 
   >>> y = physics.A(x) # now, the filter is well defined and this line does the same as above 
   >>> y = physics.A(x, filter=torch.rand((1,1,2,2))) # we set and apply a random blur filter 
   >>> y = physics.A(x) # the convolution filter is now random 
```

File: deepinv/physics/functional/convolution.py [hunk 29]
```diff
@@ -0,0 +1,358 @@
...
+
+def conv_transpose3d(y: Tensor, filter: Tensor, padding: str = "valid"):
+    r"""
+    A helper function to perform 3D transpose convolution.
+    """
+    pass
+
+
+# %%
+# if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (minh-hai-nguyen-712) [reply deepinv/physics/functional/convolution.py:249]: I made it become minimal. The last lines of code was for benchmarking the convolution between conv2d and fft so I want to leave it here...

File: deepinv/physics/functional/hist.py (hunk 2 above)
Comment (minh-hai-nguyen-712) [reply deepinv/physics/functional/hist.py:231]: It was just for some benchmarks but yes we can remove it.

File: deepinv/physics/generator/base.py (hunk 9 above)
Comment (minh-hai-nguyen-712) [reply deepinv/physics/generator/base.py:11]: added!

File: deepinv/physics/generator/blur.py (hunk 18 above)
Comment (minh-hai-nguyen-712) [reply deepinv/physics/generator/blur.py:605]: removed

File: deepinv/tests/test_generators.py [hunk 30]
```diff
@@ -0,0 +1,874 @@
...
+    assert list(params.keys()) == keys
+
+    if "filter" in params.keys():
+        assert params["filter"].shape == (batch_size, num_channels, size[0], size[1])
+
+    if "mask" in params.keys():
+        assert params["mask"].shape == (batch_size, 2, size[0], size[1])
+
+
+@pytest.mark.parametrize("name", GENERATORS)
```
[the remark this answers is not in the document]
Comment (fsarron) [reply deepinv/tests/test_generators.py:88]: We do test the generation capabilities in the other test function `test_generation()`. We do this by fixing the seed manually and checking the generator gives the expected tensor. 

I like your suggestion of checking that the params change after a call to the `step` function, I will have a look at it.

File: deepinv/physics/generator/blur.py (hunk 14 above)
Comment (minh-hai-nguyen-712) [reply deepinv/physics/generator/blur.py:314]: added

File: deepinv/tests/test_generators.py [hunk 31]
```diff
@@ -0,0 +1,874 @@
...
+    generator, size, _ = find_generator(name, size, 1, device)
+    batch_size = 1
+    torch.manual_seed(0)
+    torch.cuda.manual_seed(0)
+    params = generator.step(batch_size=batch_size)
+
+    if name == "MotionBlurGenerator":
+        w = params["filter"]
+        if device.type == "cpu":
+            wref = torch.tensor(
```
[the remark this answers is not in the document]
Comment (fsarron) [reply deepinv/tests/test_generators.py:103]: We have found that fixing the random behaviour through the call to `manual_seed` gives different tensors whether we run on cpu or cuda. But results are consistent on a given device. 

Our solution was to explicitly test the generated tensor in a device dependent way.

File: deepinv/tests/test_generators.py [hunk 32]
```diff
@@ -0,0 +1,874 @@
...
+        w = params["sigma"]
+        if device.type == "cpu":
+            wref = torch.tensor([0.2531657219])
+        elif device.type == "cuda":
+            wref = torch.tensor([0.2055327892]).to(device)
+
+    assert torch.allclose(w, wref, atol=1e-6)
+
+
+# @pytest.mark.parametrize("name_tuple, size", list_names_shape_mix)
```
[the remark this answers is not in the document]
Comment (fsarron) [reply deepinv/tests/test_generators.py:336]: Yes. Done.

File: deepinv/tests/test_physics.py [hunk 33]
```diff
@@ -167,12 +168,42 @@ def test_operators_adjointness(name, device):
...
+
+    # Test adjointess of the adjoint function given by auto diff
+    f = adjoint_function(physics.A, x.shape, x.device)
+    y = physics.A(x)
+    error2 = (f(y) - physics.A_adjoint(y)).flatten().mean().abs()
+    assert error2 < 1e-3
+
+    # Test operator norm
+    # norm = physics.compute_norm(x)
+    # assert torch.abs(norm - norm_ref) < 0.2
```
[the remark this answers is not in the document]
Comment (fsarron) [reply deepinv/tests/test_physics.py:198]: It was for review, I think it can be deleted. To quote Pierre (on the Idris Hackaton Slack): 
> In general spectral norms of convolution operators are different from 1 (except for filters on the simplex with periodic boundary conditions)...

So we chose not to test the norm for all boundary conditions.

File: deepinv/tests/test_physics.py [hunk 34]
```diff
@@ -413,12 +439,14 @@ def test_blur(device):
         img_size=(1, x.shape[-2], x.shape[-1]),
         filter=h,
         device=device,
+        padding="circular",
```
[the remark this answers is not in the document]
Comment (fsarron) [reply deepinv/tests/test_physics.py:442]: I think it was unintended. Actually, it would be best to test this for all boundary conditions via a  a `pytest.mark` decorator. I will look at it.

File: deepinv/tests/test_generators.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (fsarron) [reply deepinv/tests/test_generators.py:74]: I added a `test_generation_newparams` function that calls step twice, and checked that the output of the 2 calls are different.

File: deepinv/tests/test_physics.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (fsarron) [reply deepinv/tests/test_physics.py:442]: BlurFFT is only implemented for circular padding, so comparison only make sense for circular padding. I removed the `padding='circular'` from the call to BlurFFT as it does not do anything effectively

File: TODO_Mambo.md
[the remark this answers is not in the document]
Comment (fsarron) [reply TODO_Mambo.md:1]: Yes !

[the remark this answers is not in the document]
Comment (fsarron) [reply TODO_Mambo.md:1]: @matthieutrs Can you do it in the next commit / push ? Thanks

File: README.rst [hunk 35]
```diff
@@ -54,22 +54,22 @@ Try out the following plug-and-play image inpainting example:
...
-                                       noise_model=dinv.physics.GaussianNoise(sigma=0.01))
-
-    data_fidelity = dinv.optim.data_fidelity.L2()
-    prior = dinv.optim.prior.PnP(denoiser=dinv.models.MedianFilter())
-    model = dinv.optim.optim_builder(iteration="HQS", prior=prior, data_fidelity=data_fidelity, \
-                                     params_algo={"stepsize": 1.0, "g_param": 0.1})
-    y = physics(x)
-    x_hat = model(y, physics)
-    dinv.utils.plot([x, y, x_hat], ["signal", "measurement", "estimate"], rescale_mode='clip')
+    >>> import deepinv as dinv
```
[the remark this answers is not in the document]
Comment (fsarron) [reply README.rst:57]: We copied the style of the remaining of the docs. You can remove it or leave it, as you wish.

File: deepinv/tests/test_physics.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (matthieutrs) [reply deepinv/tests/test_physics.py:438]: Yes but here I wonder if it is meaningful to test the noise domain in the case of circular padding?

File: deepinv/physics/blur.py [hunk 36]
```diff
@@ -534,51 +403,82 @@ def V(self, x):
...
-#
-#     plt.imshow(x.squeeze(0).permute(1, 2, 0).cpu().numpy())
-#     plt.show()
-#     plt.imshow(y.squeeze(0).permute(1, 2, 0).cpu().numpy())
-#     plt.show()
-#     plt.imshow(xhat.squeeze(0).permute(1, 2, 0).cpu().numpy())
-#     plt.show()
-#     plt.imshow(xhat2.squeeze(0).permute(1, 2, 0).cpu().numpy())
-#     plt.show()
+#     import deepinv as dinv
```
Comment (matthieutrs) [inline deepinv/physics/blur.py:473]: can we remove these?

**Review by matthieutrs — CHANGES_REQUESTED** [review]

**Comment by matthieutrs** [comment]
Ok, eventually finished a first round of review! Up to very minor stuff two things are still missing:
1. testing the `step()` method for the generator; (@fsarron still ok to look into this?) 
2. fix the pseudo-inverse of the pan-sharpening operator in the case of gaussian kernel (when the kernel is None things work well). There I am a bit stuck, maybe @tachella knows better!

**Comment by fsarron** [comment]
> Ok, eventually finished a first round of review! Up to very minor stuff two things are still missing:
> 
> 1. testing the `step()` method for the generator; (@fsarron still ok to look into this?)
> 2. fix the pseudo-inverse of the pan-sharpening operator in the case of gaussian kernel (when the kernel is None things work well). There I am a bit stuck, maybe @tachella knows better!

@matthieutrs I did implement the tests for the `step()` function already, in the function `test_generation_newparams` of [deepinv/tests/test_generators.py#L75](https://github.com/deepinv/deepinv/blob/abc32e8c9d9869531665493bff7af1b0c12f5e2e/deepinv/tests/test_generators.py#L75). 

I check that upon two successive `step()` calls, at least one of the values of the param tensor have changed. I used `torch.any()` rather than `torch.all()` in case some values stay the same (as would happen for e.g. `AccelerationMaskGenerator` given it is a binary mask). 

I referenced this in a earlier review comment that I left open. I let you review it.

**Review by fsarron — APPROVED** [review]
